File: app/src/main/java/cn/ifafu/ifafu/util/GlobalLib.java
package cn.ifafu.ifafu.util;

import android.app.Activity;
import android.content.Context;
import android.content.ContextWrapper;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.view.View;
import android.view.inputmethod.InputMethodManager;

import java.util.List;

import cn.ifafu.ifafu.entity.Score;

public class GlobalLib {

    public static String trimZero(String s) {
        if (s.indexOf(".") > 0) {
            // 去掉多余的0
            s = s.replaceAll("0+?$", "");
            // 如最后一位是.则去掉
            s = s.replaceAll("[.]$", "");
        }
        return s;
    }

    public static String formatFloat(float num, int digit) {
        return trimZero(String.format("%."+ digit + "f", num));
    }

    public static float getIES(List<Score> scoreList) {
        if (scoreList == null || scoreList.isEmpty()) {
            return 0F;
        }
        float totalScore = 0;
        float totalCredit = 0;
        float totalMinus = 0;
        for (Score score : scoreList) {
            if (score.isIESItem()) {
                float calcScore = score.getCalcScore();
                totalScore += calcScore;
                totalCredit += score.getCredit();
                if (score.getRealScore() < 60) {
                    totalMinus -= score.getCredit();
                }
            }
        }
        return totalScore / totalCredit - totalMinus;
    }

    public static String getLocalVersionName(Context context) {
        String localVersion = "";
        try {
            PackageInfo packageInfo = context.getApplicationContext()
                    .getPackageManager()
                    .getPackageInfo(context.getPackageName(), 0);
            localVersion = packageInfo.versionName;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return localVersion;
    }

    public static int getLocalVersionCode(Context context) {
        try {
            PackageInfo packageInfo = context.getApplicationContext()
                    .getPackageManager()
                    .getPackageInfo(context.getPackageName(), 0);
            return packageInfo.versionCode;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
            return 9999;
        }
    }

    public static Activity getActivityFromView(View view) {
        if (null != view) {
            Context context = view.getContext();
            while (context instanceof ContextWrapper) {
                if (context instanceof Activity) {
                    return (Activity) context;
                }
                context = ((ContextWrapper) context).getBaseContext();
            }
        }
        return null;
    }

    public static void hideSoftKeyboard(Activity activity) {
        InputMethodManager imm = (InputMethodManager) activity.getSystemService(Context.INPUT_METHOD_SERVICE);
        if (imm != null) {
            imm.hideSoftInputFromWindow(activity.getWindow().getDecorView().getWindowToken(), 0);
        }
    }
}

File: app/src/main/java/cn/ifafu/ifafu/util/GlobalLib.kt
package cn.ifafu.ifafu.util

import android.app.Activity
import android.content.ContextWrapper
import android.view.View
import cn.ifafu.ifafu.data.entity.Score

object GlobalLib {
    fun trimZero(s: String): String {
        var s = s
        if (s.indexOf(".") > 0) { // 去掉多余的0
            s = s.replace("0+?$".toRegex(), "")
            // 如最后一位是.则去掉
            s = s.replace("[.]$".toRegex(), "")
        }
        return s
    }

    fun formatFloat(num: Float, digit: Int): String {
        return trimZero(String.format("%." + digit + "f", num))
    }

    @JvmStatic
    fun getActivityFromView(view: View?): Activity? {
        if (null != view) {
            var context = view.context
            while (context is ContextWrapper) {
                if (context is Activity) {
                    return context
                }
                context = context.baseContext
            }
        }
        return null
    }

    fun List<Score>.calcIES(): Float {
        // 过滤 不计入智育分的成绩 和 免修成绩
        val scores = this
                .filter { it.isIESItem && it.realScore != Score.FREE_COURSE }
        if (scores.isEmpty()) {
            return 0f
        }
        var totalScore = 0f
        var totalCredit = 0f
        var totalMinus = 0f
        for (score in scores) {
            val realScore = score.realScore
            totalScore += (realScore * score.credit)
            totalCredit += score.credit
            if (realScore < 60) {
                totalMinus += score.credit
            }
        }
        var result = totalScore / totalCredit - totalMinus
        if (result.isNaN()) {
            result = 0F
        }
        return result
    }
}